File: lib/src/main/java/net/kibotu/android/recyclerviewpresenter/UIDGenerator.java
package net.kibotu.android.recyclerviewpresenter;

class UIDGenerator {

    public static final int START_UID = 0;
    public static final int INVALID_UID = START_UID - 1;

    private static final SynchronizedValue<Integer> nextUID;

    static {
        nextUID = new SynchronizedValue<>();
        nextUID.set(0);
    }

    public static int newUID() {
        if (!isValid(nextUID.get())) {
            throw new IllegalStateException("UID pool depleted");
        }
        nextUID.set(nextUID.get() + 1);
        return nextUID.get();
    }

    private static boolean isValid(final int uid) {
        return uid >= START_UID;
    }
}

File: lib/src/main/java/net/kibotu/android/recyclerviewpresenter/UIDGenerator.kt
package net.kibotu.android.recyclerviewpresenter

import java.util.concurrent.atomic.AtomicInteger

/**
 * Created by [Jan Rabe](https://kibotu.net).
 */
internal object UIDGenerator {

    private const val START_UID = 0

    /**
     * https://winterbe.com/posts/2015/05/22/java8-concurrency-tutorial-atomic-concurrent-map-examples/
     */
    private val nextUID = AtomicInteger(START_UID)

    fun newUID(): Int {
        if (!isValid(nextUID.get())) {
            throw IllegalStateException("UID pool depleted")
        }
        return nextUID.incrementAndGet()
    }

    private fun isValid(uid: Int): Boolean = uid >= START_UID
}